### ISSUE TITLE
Introduce a process for new features

### ISSUE THREAD

**Issue by Fohlen** [inexorgame/entity-component-system#109]
We talked a lot about how we can use processes to make our life easier.

Now is the time we should tackle one of our first processes: the feature process.

For new features it would be good to have such a procedure in order to be able to introduce only high-quality, well-designed and fit features (modules).

@a-teammate and I already talked a lot about how this could work and I will present our solution below:

- a feature is drafted (e.g on HackMD)
- then an issue is opened to keep track
- after the issue (or while it is open) an implementation phase follows in a dedicated branch
- *if* the module introduces new entity types, an additional PR is created for the entity system wiki on the website (much like an RFC)
- the wiki PR is merged once finalised and then the feature branch is merged
- `=> new feature in master`

Input and ideas how this could go smoothly with the way we work is appreciated!

**Comment by a-teammate** [comment]
### Sidenote:
I think the main purpose is to make **master** always the reviewed, documented and tested state, while enforcing **branch** (in git, or informal in a pad) the steps to merge it are clear and optimally small.

#### Carefully!
There are also issues with this approach, basically since it is a formalized procedure, this procedure must
 
1. not introduce too much overhead, so everyone can adopt it
    * not too much context switches
    * not too much waiting for other persons or CI
    * not too different from what how things were approached before
2. give people benefit they see, so people want to adopt it

For all cases from the iterative "I will quickly try *this*" .. "and *this*" .. "and *this*" .. Hanack has mastered to perfection to the more indirect "ive a rough understanding about *this*? what is *this* more concretely? why do I want *this*? what can I use for *this* from what I have?" we do in hackpads.

Both approaches need to work with that formal procedure.
So since it is hard to think about new feature-names, we should either track "feature-xy-4" things or at least let them stay possible.

If we slow down adding features by having too much overhead for it for the acting person, there is no win at all in this. Doing sth "almost right" is most often better than not doing sth.

**Comment by IAmNotHanni** [comment]
As @a-teammate pointed out, it would be nice to have a set of questions about the planned feature which the developer has to answer in a hackpad for example.

* What is the name of the feature?
* Will this feature remove other features?
* What are the benefits of this feature compared with other features?
* Which additional dependencies will be introduces with this feature?

and so on..

**Comment by IAmNotHanni** [comment]
Need feedback:

![BPMN_diagram_for_a_new_feature_in_Inexor](https://user-images.githubusercontent.com/7914428/55430876-713cb900-558f-11e9-8023-f217e62c1311.png)

**Comment by IAmNotHanni** [comment]
@Fohlen suggested that we document the code in our wiki before we accept the pull request. This may be done by the maintainers and not by the contributers.

**Comment by Fohlen** [comment]
@IAmNotHanni what's with the documentation? Can you add it to the BPMN as an optional thing?
Once that's added this can be closed with a follow-up to implement (mostly document) this in the entity-system repo.
